src/services/token.js

```diff
@@ -0,0 +1,196 @@
+    },
+
+    /**
+     * Verify a JWT.
+     *
+     * @param {string} token The token to be verified.
+     * @returns {Promise}
+     */
+    verify: function (token) {
+      var config = require("../config");
+      var result = jwt.verify(token, config.jwtSecret);
+
+      if (result) {
+        return Promise.resolve(result);
+      } else {
+        return Promise.reject();
+      }
+    }
+  };
+}
+
+function SqliteMethod() {
+  // Use the None method as a "naive" implementation so that we can re-use
+  // basic JWT handling logic.
+  var naive = NoneMethod();
+
+  return {
+    generate: function (grant) {
+      var generatedToken;
+
+      // First generate a JWT using the naive implementation
+      return naive.generate(grant)
+      // Verifying it is an easy way to generate initialized and expiration times
+      .then(function (token) {
+        generatedToken = token;
+
+        return naive.verify(token.access_token);
+      })
+      // Store the generated token in the database
+      .then(function (verified) {
+        var config = require("../config");
+        var db = config.databaseService;
+
+        return Promise.all([
+          db.run(
+            "INSERT INTO refresh_token VALUES (?, ?)",
+            generatedToken.client_id,
+            generatedToken.refresh_token
+          ),
+          db.run(
+            "INSERT INTO access_token VALUES (?, ?, ?)",
+            generatedToken.access_token,
+            verified.iat,
+            verified.exp
+          )
+        ]);
+      })
+      // Calling code expects a promise with a generated token, not the output
+      // of a SQL insert.
+      .then(function () {
+        return generatedToken;
+      });
+    },
+
+    revoke: function (token, tokenType) {
+      var config = require("../config");
+      var db = config.databaseService;
+
+      // The OAUTH spec is a bit weird for revokes.
+      // > If the server is unable to locate the token using
+      // > the given hint, it MUST extend its search across all of its
+      // > supported token types.
+      // So we check for the token in both tables, and order the returned
+      // results by which one the client provided.
+      var query = [
+        "SELECT 'access_token' as token_type, 'access_token' = $token_type as is_token_type",
+        "FROM access_token WHERE access_token = $token",
+        "UNION",
+        "SELECT 'refresh_token' as token_type, 'refresh_token' = $token_type as is_token_type",
+        "FROM refresh_token WHERE refresh_token = $token",
+        "ORDER BY is_token_type",
+        "LIMIT 1"
+      ].join("\n");
+
+      var params = {
+        "$token": token,
+        "$token_type": tokenType
+      };
+
+      // Delete the token if we've got it
+      return db.get(query, params)
+      .then(function (row) {
+        var query;
+
+        switch (row["token_type"]) {
+        case "access_token":
+          query = "DELETE FROM access_token WHERE access_token = ?";
```

Comment:
I wonder if we should update the rows by setting revokedAt=now(), or something similar, rather than deleting. How much chaos would that cause?